### ISSUE TITLE
Potential API change for groups and collections

### ISSUE THREAD

**Issue by patrickarlt** [patrickarlt/acetate#31]
@ngoldman @paulcpederson @nikolaswise

In the process of documenting groups and collections it actually looks really weird and magical.

Consider if I put this in my config.

```js
acetate.group('documentation', 'documentation/*', {
  groupBy: 'topic',
  sortGroups: function(group){
    var order = ['Pages', 'Configuration', 'Tools & Plugins', 'Misc.'];
    return order.indexOf(group.name);
  },
  sortPages: 'order'
});
```

This would make a magical local variable called `documentation` that is in all my pages that match the `documentation/*` matcher. So I can use it in the templates like.

```html
{% for group in documentation.groups %}
  <section>
    <h4>{{ group.name }}</h4>
    <ul>
      {% for page in group.pages %}
      <li><a href="{{relativeUrl}}{{page.url}}">{{page.title}}</a></li>
      {% endfor %}
    </ul>
  </section>
{% endfor %}
```

But when I want to use the 'documentation' collection somewhere (that might not be in `documentation/*'`) else I need to apply this metadata.

```yaml
groups: 
  documentation: documentation
```

or 

```js
  acetate.metadata('somepage.html', {
    groups: {
      documentation: 'documentation.'
    }
  });
```

This is weird and really hard to explain in the docs. So the change I'm proposing is this.

If I want to get a group or collection on a page there is a global `collection` helper that will fetch a collection. So the above example could be written as...

```html
{% for group in group('documentation') %}
  <!-- Do Stuff-->
{% endfor %}

{% for group in collection('posts') %}
  <!-- Do Stuff-->
{% endfor %}
```

----

The second part of this is that on each page in a group or collection get a sequence so you can build say a list of blog posts or have a next link at the bottom of a tutorial. This is also unclear because it ends up looking like...

```html
{% if posts.previous %}
  <a href="{{relativeUrl}}{{ posts.previous.url }}">{{ posts.previous.title }}</a>
{% endif%}

{% if posts.next %}
  <a href="{{relativeUrl}}{{ posts.next.url }}">{{ posts.next.title }}</a>
{% endif %}
```

I was thinking these could be replaced with similar `next` and `previous` helpers so the above sample could be written like this.

```html
{% if previous('posts') %}
  <a href="{{relativeUrl}}{{ previous('posts').previous.url }}">{{ previous('posts').previous.title }}</a>
{% endif%}

{% if next('posts') %}
  <a href="{{relativeUrl}}{{ next('posts').previous.url }}">{{ next('posts').previous.title }}</a>
{% endif%}
```

**Comment by patrickarlt** [comment]
This might make watching a page with a particular collection (like a homepage with blog posts) for changes harder.

**Comment by patrickarlt** [comment]
Basically wondering if the [Collections](http://acetate.io/documentation/collections/) and [Groups](http://acetate.io/documentation/collections/) are to hard to follow

**Comment by patrickarlt** [comment]
Would namespacing all of this make it easier? for example groups/collections/data would be arrays that would transform into objects?

```html
---
data:
  - gists.js
collections:
  - blog
groups: 
  - documentation
---

{% for group in groups.documentation %}
  <!-- Do Stuff-->
{% endfor %}

{% for post in collections.blog %}
  <!-- Do Stuff-->
{% endfor %}

{% for gist in data.gists %}
  <!-- Do Stuff-->
{% endfor %}
```

**Comment by paulcpederson** [comment]
@patrickarlt I need to read over the doc for collections vs. groups because I still don't really understand the difference.

**Comment by ngoldman** [comment]
Agree, it's not very clear to me why there needs to be a distinction or what purpose the distinction serves.

**Comment by patrickarlt** [comment]
@paulcpederson @ngoldman The point of collections and groups is to create useful sets of pages for things like navigation structures, blogs ect...

* Collections - query a set of pages and order then adding a reference to the next and previous page to each page. **Blog**
* Groups - query a set of pages and group the matches by a specific metadata key, then order all the pages adding a reference to the next and previous page to each page. **One level deep navs**

The idea was to provide some way to quickly query sets of pages and add some additional behavior (like binding prev/next pages) for some common use cases.

**Comment by paulcpederson** [comment]
so groups returns multiple arrays, one for each metadata key?

**Comment by patrickarlt** [comment]
Collection looks like this

```js
[
  page,
  page,
  page
]
```

Group looks like

```
[
  {
    name: 'Shared Key'
    slug: 'shared-key',
    pages: [
      page,
      page,
      page
    ]
  },
  {
    name: 'Other Key'
    pages: [
      page,
      page,
      page
    ]
  }
]
```

**Comment by paulcpederson** [comment]
Ok. I think I'm getting this. 

And groups and collections are available globally to the whole site, right?

**Comment by paulcpederson** [comment]
I mean, it would probably be best to be explicit in importing it in the page's frontmatter. IE:

```
collections:
  - blog
```

That way you know what you're getting in every template and the watcher has a way to flag what pages need to update when a collection updates.

**Comment by paulcpederson** [comment]
Is it possible to use dot notation for javascript objects in the yaml frontmatter? or is all of that parsed as strings? I mean something like:

```
blogs: collections.blog
```

Then you could name the collection whatever you want locally in your template.

**Comment by paulcpederson** [comment]
I do think that

```
groups:
  - documentation
```

is much clearer then 

```
groups: 
  documentation: documentation
```


so I guess my vote is for: 

```
---
data:
  - gists.js
collections:
  - blog
groups: 
  - documentation
---

{% for group in groups.documentation %}
  <!-- Do Stuff-->
{% endfor %}

{% for post in collections.blog %}
  <!-- Do Stuff-->
{% endfor %}

{% for gist in data.gists %}
  <!-- Do Stuff-->
{% endfor %}
```

because it would make accessing the data the same on every page and it is very clear what is going on.

**Comment by patrickarlt** [comment]
Making data match groups and collections means that you cant do this because all we can do is match the filename. So if we do that route you can't do this.

```yaml
---
data:
  - gists.js
  - folder/gists.js
---
```

Or

```yaml
---
data:
  - gists.js
  - folder/gists.json
---
```

**Comment by patrickarlt** [comment]
I also cooked up an abstract `query` helper just to see if i could.

```js
// create a query called 'blog' that has a certian value
acetate.query('blog', function(pages){
  return _(pages).filter(function(page){
    return minimatch(page.src, 'blog/**/*');
  }).sortByOrder(['date'], [false]);
});
```

Then in our templates

```html
<h3>All Posts</h3>

<ul>
{% for post in query('blog') %}
  <li><a href="{{post.relativeUrl}}{{post.url}}">{{post.title}}</a></li>
{% endfor %}
</ul>
```

I also figured out a way to detect pages that use `query('blog')` and rebuild then when anything in `blog` query changes.

**Comment by patrickarlt** [comment]
@paulcpederson I just remembered why I didn't go the array route for data/collections/groups.

Given this page...

```html
---
data:
  - gists.js
collections:
  - blog
groups: 
  - documentation
---
```

and this config...

```js
acetate.metadata('**/*', {
  data: ['people.json']
});
```

Would totally override the existing value of data. Making so we no longer got `gists.js` on the page. Objects merge cleanly and give you the ability to change the variable name for the template which may be useful.

This could be solved by a `acetate.data(pattern, dataFilesArray)` helper that would append vs. override making it possible to safely add data references to lots of pages.

**Comment by paulcpederson** [comment]
@patrickarlt you know, call me crazy, but you could *really* abstract it and just cover all of these cases with data. 

Just call `.js` data with `site, page, callback` so that people have access to the site in their data file:

```
module.exports = function(site, page, callback){
  someData = ['thing1', 'thing2']
  callback(null, somedata);
}
```

`site` is just a sitemap, which I assume you already create at some point. To create a collection, just find the right pages from the sitemap:

```
var _ = require('lodash')
var minimatch = require('minimatch')

module.exports(site, page, callback){
  return _(site).filter(function (p) {
    return minimatch(p.src, 'blog/**/*');
  }).sortByOrder(['date'], [false]);
});
```


Then, in templates, you always just require data by name, and that data file is only run on pages that require that data:

```
---
data:
  posts: posts.js
---
```

I would think this would make watching easier too, because everything is just a data file and you have a definitive record of what data every page needs. For example, lets say you have three pages:

```
one.html
two.html
three.html
```

only `three.html` indicates it needs `posts.js`. Now you know that you only need to rebuild two and three if that file or their layout changes. Just an idea.

**Comment by paulcpederson** [comment]
people could write an `acetate-collection` module and you could use that like this in `acetate.conf.js`:

```js
var acetateCollection = require('acetate-collection')

module.exports = function (acetate) {
  acetate.metadata('posts/**/*', {
    posts: acetateCollection('posts/**/*')
  })
}
```

**Comment by paulcpederson** [comment]
This is more abstract and composable. It would also let you delete large swaths of code.

**Comment by paulcpederson** [comment]
Plus it's less doc to write :wink:

**Comment by paulcpederson** [comment]
If you want to use `acetate-collection` without injecting metadata in your `acetate.conf.js` you could just make a data file like `my-collection`:

```js
var AcetateCollection = require('acetate-collection')
module.exports = AcetateCollection('posts/**/*')
```

So essentially, `acetate-collection` would look like this:

```js
var _ = require('lodash')
var minimatch = require('minimatch')

module.exports = function (glob) {
  return function (site, page, callback) {
      return _(site).filter(function (p) {
        return minimatch(p.src, glob)
      })
}
```

**Comment by patrickarlt** [comment]
I think i've figured a few things out from the conversation. 

1. Groups vs Collections are to hard to understand.
2. Abstract helpers are better then concrete ones.

It's true you can do all this in a data file or an extension but I really like the idea of have SOME helpers in the config file.

I've got some ideas right now let me flesh them out.

**Comment by paulcpederson** [comment]
Yeah, I didn't understand groups or collections until you pasted sample responses. So definitely include that in the doc pretty prominently if they stick around.

**Comment by ngoldman** [comment]
I might be just totally insane here but what about just using tags + data? I can see a use case for [collections](http://jekyllrb.com/docs/collections/), but collections and groups seems like too much overlap.

Another thing to consider here is that you're adding features for the sake of adding features. No one's asking for these things. I think the priority should be on stabilizing the API and completing documentation so people can use it.

**Comment by patrickarlt** [comment]
@paulcpederson @ngoldman 

> Yeah, I didn't understand groups or collections until you pasted sample responses. So definitely include that in the doc pretty prominently if they stick around.

This means that groups/collections are to confusing and I'm scrapping them in favor of more abstract helper methods.

> Another thing to consider here is that you're adding features for the sake of adding features. No one's asking for these things.

I definitely want these things. I did collections a stepping stone to a personal blog which needs ordered lists of pages. I did groups because I wanted to be able to stop statically coding navigation and dynamically build navigation structures like https://developers.arcgis.com/geotrigger-service/guide/, http://acetate.io/documentation/ and http://terraformer.io/core/.

Groups and Collections are also powering the index pages on our design comps for various projects.

The inability to treat my pages as a database has always bugged me about static site generators and is something that would make life MUCH simpler.

> I think the priority should be on stabilizing the API and completing documentation so people can use it.

I'm doing this now BECAUSE I want to stabilize the v1 API, test it and document it. I don't want to end up deprecating a bumping majors all the time because I didn't make the right API choices. I just don't feel Acetate would be ready without something for easily manipulating pages.

---

@paulcpederson

I made a change recently to pass a reference to the current page into data files. So things like this are now possible:

```js
var request = require('request');

module.exports = function(callback, page){
  request({
    method: 'GET',
    url: 'https://api.github.com/users/' + page.username + '/gists',
    json: true,
    headers: {
      'User-Agent': 'request'
    }
  }, function(err, resp, body){
    callback(err, body);
  });
};
```

I did this so you could do things like build profile page (like our team page) but dynamically get content for each user (assuming each user had their own page)

So its a pretty easy step to imagine passing the whole Acetate instance down.

```js
module.exports = function(acetate, page, callback){
  var result = _(pages).groupBy('topic').transform(function(result, pages, key){
    result.push(_.sortByOrder(pages, 'order', [true]));
  }).value();

  callback(error, result);
};
```

This would be a breaking change because i would want to move `callback` to the last argument (to conform with Node). But overall I like where this is going. But again the hard part comes from needing to watch files. If a create a subset of pages and want to rebuild any page that uses `blog.js` when I create a new post I can't do that because while I know which pages use `blog.js` I don't know WHICH pages should trigger rebuilds (`blog/posts/*`).

So while it should be possible to do this because I can see wanting config values from acetate in your data files it ultimately won't work because the changes can't be watched.

---


So with that in mind I think the direction I want to go in is similar to my `acetate.query()` helper.

Remember the following requirements, this new system should:

1. clear and explicit about what is going on. No magic
2. be suitably abstract to do pretty much anything
3. cover use cases from collection and groups
4. be easily accessible to extensions
5. should be able to be watched and rebuilt

This example replicates the functionality from a collection (blog) and a group (docs).

```js
var _ = require('lodash');

module.exports = function(acetate){

  // acetate.transform(glob, transformer);
  // 
  // modify the subset of pages matching blog/posts/*
  // * no callback so it has to be sync
  // * in place modification so it doesn't need to return anything
  acetate.transform('blog/posts/*', function(pages){
    // sort pages by data and add nextPost and previousPost to them
    _(pages).sortByOrder(['date'], [false]).each(function(page, i, pages){
      page.previousPost = pages[i - 1];
      page.nextPost = pages[i - 1];
    });
  });

  // acetate.query(name, glob, builder);
  // 
  // get the matched pages and return a value that will accessible in templates
  acetate.query('posts', 'blog/posts/*', function(pages){
    // Create a list of pages and order them by date
    // Then use {{query('blog')}} in your templates to get the results
    return _(pages).sortByOrder(['date'], [false]);
  });

  acetate.query('docs', 'documentation/*', function(pages){
    // Create an array of arrays contianing pages grouped by topic and sorted by order
    // Then use {{query('docs')}} in your templates to get the results
    return _(pages).groupBy('topic').transform(function(result, pages, key){
      result.push(_.sortByOrder(pages, 'order', [true]));
    });
  });
};
```

`acetate.transform()` will become pretty much the `git rebase` of Acetate as almost everything can map to it under the hood. It's very similar to `acetate.metadata()`, but is a bit more general.

`acetate.query()` makes it really easy to slice and dice sets of pages. The `query('name')` helper is bound to each page individually and will register that page as using a collection for the watcher to pick up on. I've tested this and it works.

The other advantage to exposing named queries is that extensions could register queries for users. For example a blogging extension could expose `{{ query('posts') }}` for people.

---

So I want to make the following changes:

1. Pass the whole site down to data files.
2. Add `acetate.query()`
2. Add `acetate.transform()`
3. Remove groups and collections

The nice part is there is a TON less doc to do and a TON less code. The `acetate.query` code is only 36 lines long right now and `acetate.transform()` will probably be around a dozen and both can be easily tested.

**Comment by paulcpederson** [comment]
@patrickarlt how do you access the data you just made available with `acetate.query('docs', 'documentation/*' ...` from a template?

**Comment by paulcpederson** [comment]
What do templates + front matter look like with `transform` and `query`

**Comment by paulcpederson** [comment]
also, I don't know if you thought of this, but now transform removes the need for pagination as you can just transform a page into multiple pages with different ranges of entries on them. (that could be an acetate plugin)

**Comment by patrickarlt** [comment]
`acetate.transform()` is just an in place modification of the page. So in the example above you could just do {{ previousPost.url }}. The whole page object is passed into the template. So my previous/next link would look like.

```js
{% if previousPost %}
  <a href="{{relativeUrl}}{{ previousPost.url }}">{{ previousPost.title }}</a>
{% endif%}

{% if nextPost %}
  <a href="{{relativeUrl}}{{ nextPost.url }}">{{ nextPost.title }}</a>
{% endif %}
```

`acetate.query()` adds a special helper to each page so you can do `{{ query('docs') }}` to get the value.

```html
<h3>All Posts</h3>

<ul>
{% for post in query('blog') %}
  <li><a href="{{post.relativeUrl}}{{post.url}}">{{post.title}}</a></li>
{% endfor %}
</ul>
```

**Comment by paulcpederson** [comment]
I like transform. It seems very powerful, and I suppose if you are editing pages in place you will understand that and it won't be magic (because you wrote the code)

Could you effectively accomplish query *with* transform? For example, add an array of all the blog pages as a `blog` variable to each page in `posts/`:

```js
acetate.transform('blog/posts/*', function(pages, site){
  var blogs =  _(site)
  .filter(p => minimatch(p.src, 'blog/**/*'))
  .sortByOrder(['date'], [false])

  pages = pages.map(page => {
    page.blog = blogs
    return page
  })
});
```

**Comment by paulcpederson** [comment]
sorry for switching to es6 there half way through, it's just easier

**Comment by patrickarlt** [comment]
@paulcpederson I know thats super tempting but how will I know which pages to rebuild when something in `blog/posts/*` changes? The code you posted will work but it wont rebuild /blog/ for example when the watcher is running. This is pretty much the inverse of what your idea of using `data` was.

We have to know 2 things for the watcher.

1. What pages can trigger a rebuild
2. What pages to rebuild when a change happens.

The code for the query helper is a little magic and registers the page using {{ query('blog') }} internally and when it sees a dirty page in `'blog/posts/*'` it also goes into its internal cache and dirties all the pages it saw use {{ query('blog') )} on the last build.

**Comment by paulcpederson** [comment]
@patrickarlt what if you solved № 1 with:

```js
acetate.transform('blog/posts/*', function(pages, site){
  var blogs =  _(site)
  .filter(p => minimatch(p.src, 'blog/**/*'))
  .sortByOrder(['date'], [false])

  pages = pages.map(page => {
    page.blog = blogs
    return page
  })
  return 'blog/**/*'
});
```

Just return the glob that invalidates the cache of this transform. Or return it as an array for multiple:

```js
return ['blog/**/*', 'data/my-extra-data.json']
```

**Comment by patrickarlt** [comment]
@paulcpederson that seems a little weird to me. It's not really clear WHY I'm returning a glob/array of globs.

I just figured out a way to duplicate the exact same behavior using a variable, so it eliminates the weird function syntax for getting a query.

```js
<h3>All Posts</h3>

<ul>
{% for post in queries.blog %}
  <li><a href="{{post.relativeUrl}}{{post.url}}">{{post.title}}</a></li>
{% endfor %}
</ul>
```

**Comment by paulcpederson** [comment]
*much* clearer

**Comment by paulcpederson** [comment]
That makes it way more obvious that I'm accessing a query called blog that I set up elsewhere

**Comment by patrickarlt** [comment]
@paulcpederson now that I know how to do this for query what do you think about similar helpers for data?

```js
acetate.data('gists', `gists.js`);
````

then

```html
<ul>
{% for gist in data.gist %}
  <li><a href="{{gist.htmlUrl}}">{{gist.description}}</a></li>
{% endfor %}
</ul>
```

**Comment by patrickarlt** [comment]
Or do you still like the frontmatter based syntax? benefit of this is that data is always accessible from every template so no more `data` in frontmatter.

**Comment by paulcpederson** [comment]
I like `data.gist` a lot. Makes sense.
I think that there should still be `data` in frontmatter.

**Comment by patrickarlt** [comment]
@paulcpederson Hummm.

So if i had this config.

```js
acetate.data('gists', `gists.js`);
```

and this page

```html
---
data: 
  myGists: gists.js
---

<ul>
{% for gist in data.gist %}
  <li><a href="{{gist.htmlUrl}}">{{gist.description}}</a></li>
{% endfor %}
</ul>
```

Would I basically be redefining the result of `gists.js` to a different variable?

**Comment by paulcpederson** [comment]
Hmmm. weird. I'm thinking about definitions of data in `acetate.conf.js` as **bulk** settings. Much like you had

```js
  acetate.metadata('**/*', {
    title: 'Acetate',
    author: 'Patrick Arlt'
  });
```

Like add the data from `gist.js` to all the pages

```js
acetate.data('**/*', {
  gists: 'gist.js'
});
```

But you could not do that and do:

```
---
data:
  gists: gists.js
---
```

then in your template, either way you would do:

```
<ul>
{% for gist in data.gist %}
  <li><a href="{{gist.htmlUrl}}">{{gist.description}}</a></li>
{% endfor %}
</ul>
```

**Comment by paulcpederson** [comment]
Like data can only be declared globally by explicitly assigning it to `**/*` in the config. So in your example you wouldn't be overriding anything, you would *only* have access to `data.myGists`

**Comment by paulcpederson** [comment]
But I guess I see what you're saying, because you're making `query` available globally, so wouldn't you just use `data.thing` and remove data from front-matter.

**Comment by patrickarlt** [comment]
What I was saying is that `acetate.data('gists', `gists.js`);` is more about defining a data source. Once the source is defined you can use it on any page with.

```js
acetate.data('gists', `gists.js`); // register `gists`
```

Then on any page:

```html
<ul>
{% for gist in data.gist %}
  <li><a href="{{gist.htmlUrl}}">{{gist.description}}</a></li>
{% endfor %}
</ul>
```

I can use the same technique I used for `query` here and  rebuild only the pages that use that data when the data changes.

**Comment by paulcpederson** [comment]
Yeah. I get it. 

The use case for leaving data in the front matter then becomes: *I want to override the data*. So:

```
acetate.data('gists', 'gists.js')
```

Then I want gists to be something different on just one page:

```
---
data:
  gists: 'gists2.js'
---
```

It's a way to override data.

**Comment by paulcpederson** [comment]
but maybe that's unnecessary.

**Comment by patrickarlt** [comment]
> but maybe that's unnecessary.

Thats what im thinking, why not just register it twice?

```js
acetate.data('gists', 'gists.js');
acetate.data('gists2', 'gists2.js');
```

**Comment by paulcpederson** [comment]
I mean overriding data being unnecessary.

**Comment by ngoldman** [comment]
I think it makes sense to be able to override it in the front matter. Also means if you only really need it on one page you don't need to go the conf.js route.

**Comment by ngoldman** [comment]
simple method for single use, global method when necessary

**Comment by paulcpederson** [comment]
+1

**Comment by ngoldman** [comment]
I guess part of my thinking is ideally acetate could run out of the box without conf at all, relying on a loose set of base conventions and front matter settings. This would cover many use cases. For more complex sites the conf would become crucial, but for a simple set of pages and articles it's an unnecessary level of configuration

**Comment by patrickarlt** [comment]
@ngoldman @paulcpederson I was really hoping I wasn't going to have to do this both ways but I now see how useful it would be.

@ngoldman should I be able to run acetate without a config file if I don't need one? For example if I cant find a config file should it just proceed with the default options? That would be pretty cool.

**Comment by paulcpederson** [comment]
yes, absolutely!

**Comment by patrickarlt** [comment]
@paulcpederson @ngoldman looking at this now its going to be tough to load global data + local data and handle local overriding global when specified and watching global + local data.

I'll mess around with it a little more today but right now i just have 2 separate data loading extensions which works except i get double builds when i both locally and globally declare a data file.

**Comment by patrickarlt** [comment]
@paulcpederson @ngoldman Overriding global data variables from locals if the source is the same is proving extremely difficult. To simplify this a bit I think that if you define a data variable in your config:

```js
acetate.data('gists', 'gists.js');
```

and locally:

```html
---
data:
  myGists: 'gists.js'
---
```

You would get BOTH {{ `data.gists` }} AND {{ `data.myGists` }}.

**Comment by paulcpederson** [comment]
That isn't how you would override anyways... With the above, getting both `{{data.gists}}` and `{{data.myGists}}` is expected.

An override would be: 

```js
acetate.data('gists', 'gists.js')
```

and 

```
---
data:
  gists: 'gists2.js'
---
```

Using the **same name** to refer to **different data** is what I mean when I say override.

**Comment by ngoldman** [comment]
It's doable, just requires some careful checks. I believe in you @patrickarlt.

**Comment by ngoldman** [comment]
You can of course implement global first then add local support in a later version. Semver to the rescue

**Comment by ngoldman** [comment]
you can make issues for it and we can help you

**Comment by patrickarlt** [comment]
@ngoldman actually I think I just got it all working.

**Comment by paulcpederson** [comment]
:clap: :clap: :clap:

**Comment by ngoldman** [comment]
actually 😎

**Comment by ngoldman** [comment]
![](http://meme.loqi.me/4aar_5B_.jpg)
